### PR TITLE
fix(datasource/nuget): Encode whitespace for source and homepage URLs

### PR DESCRIPTION
**Pull request by timmkrause**
## Changes:

Encode whitespace for parsed NuGet source and homepage URLs in order to generate working Markdown links.

## Context:

image

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/markdown.ts [hunk 1]
```diff
@@ -5,19 +5,32 @@ import { regEx } from './regex';
 // Generic replacements/link-breakers
 export function sanitizeMarkdown(markdown: string): string {
   let res = markdown;
+
+  // Replace spaces in Markdown links with %20.
+  const markdownLinks = res.match(/\(https?:\/\/[^)]*\)/g);
+
+  if (markdownLinks) {
+    markdownLinks.forEach((ml) => {
+      res = res.replace(ml, ml.replace(' ', '%20'));
```
Comment (rarkins) [inline lib/util/markdown.ts:14]: Won't this only replace the first only?

**Review by rarkins — CHANGES_REQUESTED** [review]
This is not really the correct location for this purpose. `sanitizeMarkdown()` is for trying to remove backlinks and things like that. Not for general URL massaging.

File: lib/util/markdown.ts (hunk 1 above)
Comment (rarkins) [reply lib/util/markdown.ts:14]: But what about one link with two spaces?

Comment (timmkrause) [reply lib/util/markdown.ts]: Fixed.

**Comment by timmkrause** [comment]
@rarkins What do you suggest regarding the scope? Should it be scoped to PR Markdown generation, changelog generation, general URL handling of artifacts, general URL handling of upgrades, even wider? The broader the scope the less necessary it is for the participating parts because they don't have an issue with it. Could you guide me a little into the right direction where you would apply it. Or would you want to tackle this problem at all?

File: lib/util/markdown.spec.ts [hunk 2]
```diff
@@ -29,5 +30,13 @@ describe('util/markdown', () => {
     it('works', async () => {
       expect(await linkify(before, { repository: 'some/repo' })).toEqual(after);
     });
-  });
+  }),
+    describe('.sanitizeMarkdown', () => {
+      it('sanitizes URLs of Markdown links containing spaces to working links', async () => {
+        const before = `[Some.Package](https://foo.bar/E  F/A2B CD) ([source](https://foo.bar/EF/A2B CD/_git/Some.Package))`;
+        const after = `[Some.Package](https://foo.bar/E%20%20F/A2B%20CD) ([source](https://foo.bar/EF/A2B%20CD/_git/Some.Package))`;
```
Comment (viceice) [inline lib/util/markdown.spec.ts:37]:
```suggestion
        const before = `[Some.Package](https://foo.bar/E  F/A2B CD) [Some.Package](https://foo.bar/E  F/A2B CD) ([source](https://foo.bar/EF/A2B CD/_git/Some.Package))`;
        const after = `[Some.Package](https://foo.bar/E%20%20F/A2B%20CD) [Some.Package](https://foo.bar/E%20%20F/A2B%20CD) ([source](https://foo.bar/EF/A2B%20CD/_git/Some.Package))`;
```

Should also replace same link multiple times

**Comment by timmkrause** [comment]
It all seems to begin during package creation (.NET/NuGet world here).

I specify

```
    <RepositoryUrl>https://foo.bar/EF/A2B%20CD/_git/Some.Package</RepositoryUrl>
    <PackageProjectUrl>https://foo.bar/EF/A2B%20CD</PackageProjectUrl>
```

in the project file. Creating the package via `dotnet pack` URL decodes the URLs which introduces the whitespace again.

**Comment by viceice** [comment]
Ok, then we should message them on datasource layer.

**Comment by timmkrause** [comment]
I already started with adding this to

image

Just FYI.

File: lib/util/markdown.spec.ts [hunk 3]
```diff
@@ -29,5 +30,13 @@ describe('util/markdown', () => {
...
       expect(await linkify(before, { repository: 'some/repo' })).toEqual(after);
     });
-  });
+  }),
+    describe('.sanitizeMarkdown', () => {
+      it('sanitizes URLs of Markdown links containing spaces to working links', async () => {
+        const before = `[Some.Package](https://foo.bar/E  F/A2B CD) [Some.Package](https://foo.bar/E  F/A2B CD) ([source](https://foo.bar/EF/A2B CD/_git/Some.Package))`;
+        const after = `[Some.Package](https://foo.bar/E%20%20F/A2B%20CD) [Some.Package](https://foo.bar/E%20%20F/A2B%20CD) ([source](https://foo.bar/EF/A2B%20CD/_git/Some.Package))`;
+
+        expect(await sanitizeMarkdown(before)).toEqual(after);
```
Comment (viceice) [inline lib/util/markdown.spec.ts:39]:
```suggestion
        expect(sanitizeMarkdown(before)).toEqual(after);
```

File: lib/util/markdown.spec.ts [hunk 4]
```diff
@@ -29,5 +30,13 @@ describe('util/markdown', () => {
     it('works', async () => {
       expect(await linkify(before, { repository: 'some/repo' })).toEqual(after);
     });
-  });
+  }),
+    describe('.sanitizeMarkdown', () => {
+      it('sanitizes URLs of Markdown links containing spaces to working links', async () => {
```
Comment (viceice) [inline lib/util/markdown.spec.ts:35]:
```suggestion
      it('sanitizes URLs of Markdown links containing spaces to working links', () => {
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/markdown.spec.ts (hunk 4 above)
Comment (timmkrause) [reply lib/util/markdown.spec.ts]: Obsolete, right?

File: lib/util/markdown.spec.ts (hunk 3 above)
Comment (timmkrause) [reply lib/util/markdown.spec.ts]: Obsolete, right?

File: lib/util/markdown.ts [hunk 5]
```diff
@@ -5,19 +5,23 @@ import { regEx } from './regex';
 // Generic replacements/link-breakers
 export function sanitizeMarkdown(markdown: string): string {
   let res = markdown;
+
```
Comment (viceice) [inline lib/util/markdown.ts:8]: Please revert changes to this file

Comment (timmkrause) [reply lib/util/markdown.ts]: Who resolves conversations? Is there a "policy"? Conversation issuer for re-review or the one who adresses/fixes stuff?

Comment (viceice) [reply lib/util/markdown.ts]: The targeted user 😉

**Review by viceice — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**